packages/xod-project/src/patch.js

```diff
@@ -1566,3 +1568,84 @@ export const checkSpecializationMatchesAbstraction = def(
     return Either.of(specializationPatch);
   }
 );
+
+// =============================================================================
+//
+// Functions that checks could something change in the Patch
+//
+// =============================================================================
+
+// :: StrMap Node -> Set Node
+const getSetOfNodeIds = R.pipe(R.keys, a => new Set(a));
+
+/**
+ * Checks is Patch has some changes over the Nodes that
+ * could affect on generic pin deduction or marked as
+ * "deprecated", "utility" or "dead".
+ *
+ * True if have changes.
+ */
+export const haveAddedNodesOrChangedTypesOrBoundValues = def(
+  'haveAddedNodesOrChangedTypesOrBoundValues :: Patch -> Patch -> Boolean',
```

Comment:
I think the name does not scale well. You use `isColdOrForecastIsBadOrLeaveForALongTime` instead of `shouldDressCoat`. The latter will work better as it more straightforward and does not depend on the exact such of conditions that may evolve.

~So, in this case, something like `hasChangesObservableOutside` will be more meaningful IMO, if the function is indeed about it. Please, confirm.~

A proper naming will immediately show a problem. You use `haveAddedNodesOrChangedTypesOrBoundValues` for both: project browser selector and deduced pin types selector. But they are interested in different set of criterias. Should be three functions :thinking: 

- `couldChangeProjectBrowserLook` (move to `xod-client`)
- `couldAffectTypeDeductionInside`
- `couldAffectTypeDeductionOutside`